src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,441 @@
+                    futures::future::result(
+                        symlink(requested_jdk_home, jdk_home_in_workdir)
+                            .map_err(|err| format!("Error making new symlink for local execution in workdir {:?}: {:?}", &workdir_for_server2, err))
+                    )
+                } else {
+                    let maybe_existing_jdk = read_link(jdk_home_in_workdir).map_err(|e| format!("{}", e));
+                    let maybe_existing_jdk2 = maybe_existing_jdk.clone();
+                    if maybe_existing_jdk.is_err() || (maybe_existing_jdk.is_ok() && maybe_existing_jdk.unwrap() != requested_jdk_home) {
+                        let res = remove_file(jdk_home_in_workdir2)
+                            .map_err(|err| format!(
+                                "Error removing existing (but incorrect) jdk symlink. We wanted it to point to {:?}, but it pointed to {:?}. {}",
+                                &requested_jdk_home, &maybe_existing_jdk2, err
+                            ))
+                            .and_then(|_| {
+                                symlink(requested_jdk_home, jdk_home_in_workdir3)
+                                    .map_err(|err| format!("Error overwriting symlink for local execution in workdir {:?}: {:?}", &workdir_for_server2, err))
+                            });
+                        futures::future::result(res)
+                    } else {
+                        debug!("JDK home for Nailgun already exists in {:?}. Using that one.", &workdir_for_server4);
+                        futures::future::ok(())
+                    }
+                }
+        })
+        .inspect(move |_| debug!("Materialized directory {:?} before connecting to nailgun server.", &workdir_for_server3))
+        .to_boxed()
+    }
+
+    fn get_python_distribution(&self) -> String {
+        format!("{}", self.python_distribution_absolute_path.display())
+    }
+}
+
+impl super::CommandRunner for NailgunCommandRunner {
+    fn run(
+        &self,
+        req: MultiPlatformExecuteProcessRequest,
+        context: Context) -> BoxFuture<FallibleExecuteProcessResult, String> {
+
+        let nailgun_pool = self.nailgun_pool.clone();
+        let inner = self.inner.clone();
+        let python_distribution = self.get_python_distribution();
+
+        let client_req = self.extract_compatible_request(&req).unwrap();
+
+        if !client_req.is_nailgunnable {
+            trace!("The request is not nailgunnable! Short-circuiting to regular process execution");
+            return self.inner.run(req, context)
+        }
+        debug!("Running request under nailgun:\n {:#?}", &client_req);
+
+        // Separate argument lists, to form distinct EPRs for (1) starting the nailgun server and (2) running the client in it.
+        let ParsedJVMCommandLines { nailgun_args, client_args, client_main_class } = ParsedJVMCommandLines::parse_command_lines(&client_req.argv).expect("Error parsing command lines!!");
```

Comment:
Don't `.expect` - return a future error (maybe by using `try_future!`)

Reply:
Mm, yea. This is a blocker.